File: src/domain/measurement.ts
export const measurementKinds = [
  "weight",
  "body_temperature",
  "blood_pressure_systolic",
  "blood_pressure_diastolic",
  "pulse",
] as const;

export type MeasurementKind = (typeof measurementKinds)[number];

export const measurementUnits = ["kg", "celsius", "mmHg", "bpm"] as const;

export type MeasurementUnit = (typeof measurementUnits)[number];

export const measurementSources = [
  "google_fit",
  "apple_health_export",
  "mixed",
] as const;

/**
 * Exporter records the physical measurement device name. The historical
 * google_fit and apple_health_export values remain accepted API values.
 */
export type MeasurementSource = string;

export const measurementPeriods = ["morning", "evening"] as const;

export type MeasurementPeriod = (typeof measurementPeriods)[number];

export const measurementPeriodLabels = {
  morning: "朝",
  evening: "夜",
} as const satisfies Record<MeasurementPeriod, string>;

export interface MeasurementReading {
  readonly kind: MeasurementKind;
  readonly value: number;
  readonly unit: MeasurementUnit;
  readonly measuredAt: string;
  readonly source: string;
  readonly sourceRecordId?: string;
}

/**
 * F-2: counts are kept in both units and never collapsed into one (AC-57).
 * `windowedReadingCount` answers #54 (their published record count) and
 * `uniqueMeasurementCount` answers #38 and the Slice 7 gate (AC-59).
 */
/**
 * V-3 (design §7.1): the pipeline decides the transfer outcome from what the
 * sheet actually reports, not from "the call didn't throw". `written` means
 * the adapter got a success response; whether cells actually changed is a
 * separate question answered by `transferredCellCount`.
 */
export interface TransferOutcome {
  readonly state: "written" | "not-written" | "unknown";
  readonly transferredCellCount?: number;
}

export interface MeasurementCounts {
  /** Published records inside the window, after exact same-path deduplication. */
  readonly windowedReadingCount: number;
  /** Physical measurements, after cross-path identity merges them (D-5). */
  readonly uniqueMeasurementCount: number;
}

export interface LatestMeasurementSet {
  readonly period: MeasurementPeriod;
  readonly capturedAt: string;
  readonly source: MeasurementSource;
  readonly counts: MeasurementCounts;
  readonly weightKg?: number;
  readonly bodyTemperatureCelsius?: number;
  readonly bloodPressureSystolicMmHg?: number;
  readonly bloodPressureDiastolicMmHg?: number;
  readonly pulseBpm?: number;
  readonly sourcesByKind: Partial<
    Record<MeasurementKind, string>
  >;
}

export interface SpreadsheetRow {
  readonly date: string;
  readonly time: string;
  readonly periodLabel: (typeof measurementPeriodLabels)[MeasurementPeriod];
  readonly weightKg: number | "";
  readonly bodyTemperatureCelsius: number | "";
  readonly bloodPressureSystolicMmHg: number | "";
  readonly bloodPressureDiastolicMmHg: number | "";
  readonly pulseBpm: number | "";
  readonly source: MeasurementSource;
}

/**
 * E-3: the transferred value is rounded to the resolution of its kind, so no
 * path priority has to be decided and `68.19999694824219` cannot reach the
 * spreadsheet (AC-60, AC-61). Resolutions are 0.1 kg, 0.1 celsius, 1 mmHg and
 * 1 bpm, expressed as decimal places to keep the rounding exact in binary.
 */
const measurementResolutionDigits = {
  weight: 1,
  body_temperature: 1,
  blood_pressure_systolic: 0,
  blood_pressure_diastolic: 0,
  pulse: 0,
} as const satisfies Record<MeasurementKind, number>;

export function roundToMeasurementResolution(
  value: number,
  kind: MeasurementKind,
): number {
  return Number(value.toFixed(measurementResolutionDigits[kind]));
}

export function latestByKind(
  readings: readonly MeasurementReading[],
): Map<MeasurementKind, MeasurementReading> {
  const latestMap = new Map<MeasurementKind, MeasurementReading>();

  for (const reading of readings) {
    const current = latestMap.get(reading.kind);
    if (
      !current ||
      Date.parse(reading.measuredAt) > Date.parse(current.measuredAt)
    ) {
      latestMap.set(reading.kind, reading);
    }
  }

  return latestMap;
}

export function selectWeightByPeriod(
  readings: readonly MeasurementReading[],
  period: MeasurementPeriod,
): MeasurementReading | undefined {
  let selected: MeasurementReading | undefined;

  for (const reading of readings) {
    if (reading.kind !== "weight") {
      continue;
    }

    if (
      !selected ||
      shouldReplaceWeight(reading, selected, period)
    ) {
      selected = reading;
    }
  }

  return selected;
}

export function selectReadingsByWeightAnchor(
  readings: readonly MeasurementReading[],
  period: MeasurementPeriod,
): Map<MeasurementKind, MeasurementReading> {
  const selectedWeight = selectWeightByPeriod(readings, period);
  const selectedReadings = new Map<MeasurementKind, MeasurementReading>();

  if (!selectedWeight) {
    return selectedReadings;
  }

  selectedReadings.set("weight", selectedWeight);

  for (const kind of measurementKinds) {
    if (kind === "weight") {
      continue;
    }

    const closest = selectClosestToReference(
      readings,
      kind,
      selectedWeight.measuredAt,
    );
    if (closest) {
      selectedReadings.set(kind, closest);
    }
  }

  return selectedReadings;
}

function selectClosestToReference(
  readings: readonly MeasurementReading[],
  kind: MeasurementKind,
  referenceTime: string,
): MeasurementReading | undefined {
  let selected: MeasurementReading | undefined;
  const referenceMs = Date.parse(referenceTime);

  for (const reading of readings) {
    if (reading.kind !== kind) {
      continue;
    }

    if (
      !selected ||
      Math.abs(Date.parse(reading.measuredAt) - referenceMs) <
        Math.abs(Date.parse(selected.measuredAt) - referenceMs)
    ) {
      selected = reading;
    }
  }

  return selected;
}

function shouldReplaceWeight(
  candidate: MeasurementReading,
  current: MeasurementReading,
  period: MeasurementPeriod,
): boolean {
  const candidateMs = Date.parse(candidate.measuredAt);
  const currentMs = Date.parse(current.measuredAt);

  return period === "morning"
    ? candidateMs < currentMs
    : candidateMs > currentMs;
}

File: internal/domain/measurement.go
package domain

import (
	"math"
	"time"
)

type MeasurementKind string

const (
	KindWeight                 MeasurementKind = "weight"
	KindBodyTemperature        MeasurementKind = "body_temperature"
	KindBloodPressureSystolic  MeasurementKind = "blood_pressure_systolic"
	KindBloodPressureDiastolic MeasurementKind = "blood_pressure_diastolic"
	KindPulse                  MeasurementKind = "pulse"
)

var MeasurementKinds = []MeasurementKind{
	KindWeight,
	KindBodyTemperature,
	KindBloodPressureSystolic,
	KindBloodPressureDiastolic,
	KindPulse,
}

type MeasurementUnit string

const (
	UnitKg      MeasurementUnit = "kg"
	UnitCelsius MeasurementUnit = "celsius"
	UnitMmHg    MeasurementUnit = "mmHg"
	UnitBPM     MeasurementUnit = "bpm"
)

type MeasurementPeriod string

const (
	PeriodMorning MeasurementPeriod = "morning"
	PeriodEvening MeasurementPeriod = "evening"
)

var PeriodLabels = map[MeasurementPeriod]string{
	PeriodMorning: "朝",
	PeriodEvening: "夜",
}

type MeasurementReading struct {
	Kind           MeasurementKind
	Value          float64
	Unit           MeasurementUnit
	MeasuredAt     string
	Source         string
	SourceRecordID string
}

type MeasurementCounts struct {
	WindowedReadingCount   int `json:"windowedReadingCount"`
	UniqueMeasurementCount int `json:"uniqueMeasurementCount"`
}

type LatestMeasurementSet struct {
	Period                     MeasurementPeriod
	CapturedAt                 string
	Source                     string
	Counts                     MeasurementCounts
	WeightKg                   *float64
	BodyTemperatureCelsius     *float64
	BloodPressureSystolicMmHg  *float64
	BloodPressureDiastolicMmHg *float64
	PulseBpm                   *float64
	SourcesByKind              map[MeasurementKind]string
}

type SpreadsheetRow struct {
	Date                       string
	Time                       string
	PeriodLabel                string
	WeightKg                   any
	BodyTemperatureCelsius     any
	BloodPressureSystolicMmHg  any
	BloodPressureDiastolicMmHg any
	PulseBpm                   any
	Source                     string
}

func RoundToMeasurementResolution(value float64, kind MeasurementKind) float64 {
	digits := 0.0
	if kind == KindWeight || kind == KindBodyTemperature {
		digits = 1
	}
	scale := math.Pow10(int(digits))
	return math.Round(value*scale) / scale
}

func LatestByKind(readings []MeasurementReading) map[MeasurementKind]MeasurementReading {
	latest := make(map[MeasurementKind]MeasurementReading)
	for _, reading := range readings {
		current, ok := latest[reading.Kind]
		if !ok || measuredAfter(reading.MeasuredAt, current.MeasuredAt) {
			latest[reading.Kind] = reading
		}
	}
	return latest
}

func SelectWeightByPeriod(readings []MeasurementReading, period MeasurementPeriod) *MeasurementReading {
	var selected *MeasurementReading
	for i := range readings {
		reading := readings[i]
		if reading.Kind != KindWeight {
			continue
		}
		if selected == nil || shouldReplaceWeight(reading, *selected, period) {
			copy := reading
			selected = &copy
		}
	}
	return selected
}

func SelectReadingsByWeightAnchor(readings []MeasurementReading, period MeasurementPeriod) map[MeasurementKind]MeasurementReading {
	selected := make(map[MeasurementKind]MeasurementReading)
	weight := SelectWeightByPeriod(readings, period)
	if weight == nil {
		return selected
	}
	selected[KindWeight] = *weight
	for _, kind := range MeasurementKinds {
		if kind == KindWeight {
			continue
		}
		if reading, ok := selectClosestToReference(readings, kind, weight.MeasuredAt); ok {
			selected[kind] = reading
		}
	}
	return selected
}

func shouldReplaceWeight(candidate, current MeasurementReading, period MeasurementPeriod) bool {
	candidateTime, candidateOK := parseMeasuredAt(candidate.MeasuredAt)
	currentTime, currentOK := parseMeasuredAt(current.MeasuredAt)
	if !candidateOK || !currentOK {
		return false
	}
	if period == PeriodMorning {
		return candidateTime.Before(currentTime)
	}
	return candidateTime.After(currentTime)
}

func selectClosestToReference(readings []MeasurementReading, kind MeasurementKind, reference string) (MeasurementReading, bool) {
	referenceTime, referenceOK := parseMeasuredAt(reference)
	var selected MeasurementReading
	var selectedTime time.Time
	selectedOK := false
	for _, reading := range readings {
		if reading.Kind != kind {
			continue
		}
		readingTime, ok := parseMeasuredAt(reading.MeasuredAt)
		if !ok || !referenceOK {
			continue
		}
		if !selectedOK || absDuration(readingTime.Sub(referenceTime)) < absDuration(selectedTime.Sub(referenceTime)) {
			selected = reading
			selectedTime = readingTime
			selectedOK = true
		}
	}
	return selected, selectedOK
}

func measuredAfter(candidate, current string) bool {
	candidateTime, candidateOK := parseMeasuredAt(candidate)
	currentTime, currentOK := parseMeasuredAt(current)
	return candidateOK && currentOK && candidateTime.After(currentTime)
}

func parseMeasuredAt(value string) (time.Time, bool) {
	parsed, err := time.Parse(time.RFC3339Nano, value)
	return parsed, err == nil
}

func absDuration(value time.Duration) time.Duration {
	if value < 0 {
		return -value
	}
	return value
}

func CanonicalISO(value time.Time) string {
	return value.UTC().Format("2006-01-02T15:04:05.000Z")
}
